chia/full_node/mempool_manager.py

```diff
@@ -103,6 +117,84 @@ def compute_assert_height(
+
+    Args:
+        item: the mempool item to check
+        dedup_coin_spends: the current global state of identical spend aggregation
+
+    Returns:
+        List[CoinSpend]: list of coin spends to deduplicate from this mempool item
+        uint64: the cost we're saving by this deduplication
+        Dict[bytes32, DedupCoinSpend]: new deduplication information to update the global state with
+        Set[Coin]: the additions that need to be deduplicated as a result of spend deduplication
+    """
+    cost_saving = uint64(0)
+    spends_to_dedup: List[CoinSpend] = []
+    dedup_additions: Set[Coin] = set()
+    new_dedups: Dict[bytes32, DedupCoinSpend] = {}
+    # See if this item has coin spends that are eligible for deduplication
+    assert item.npc_result.conds is not None
+    eligible_coin_ids = [bytes32(s.coin_id) for s in item.npc_result.conds.spends if s.flags & ELIGIBLE_FOR_DEDUP]
+    for coin_id in eligible_coin_ids:
+        # Pick the first coin spend with coin_id
+        # npc_result.conds.spends should be consistent with
+        # spend_bundle.coin_spends at this point, guaranteeing the presence of
+        # this coin spend in this bundle as a result
+        coin_spend_in_bundle = next(cs for cs in item.spend_bundle.coin_spends if cs.coin.name() == coin_id)
+        solution_in_bundle = coin_spend_in_bundle.solution
+        # See if we processed an item with this coin before
+        if coin_id not in dedup_coin_spends:
+            # We didn't process an item with this coin before. If we end up including
+            # this item, add this pair to dedup_coin_spends
+            new_dedups[coin_id] = DedupCoinSpend(solution_in_bundle, None, set())
+            continue
+        # See if the solution was identical
+        current_solution, duplicate_cost, duplicate_additions = dataclasses.astuple(dedup_coin_spends[coin_id])
+        if current_solution != solution_in_bundle:
+            # It wasn't, so let's skip this whole item because it's relying on
+            # spending this coin with a different solution and that would
+            # conflict with the coin spends that we're deduplicating already
+            raise ValueError("Solution is different from what we're deduplicating on")
+        # Let's calculate the saved cost if we never did that before
+        if duplicate_cost is None:
+            # If we never ran this before, additions should be empty
+            assert len(duplicate_additions) == 0
+            spend_cost, created_coins = run_for_cost_and_additions(
+                coin_id, coin_spend_in_bundle.puzzle_reveal, coin_spend_in_bundle.solution, item.cost
+            )
+            duplicate_cost = spend_cost
+            duplicate_additions.update(created_coins)
+            # If we end up including this item, update this entry's cost and additions
+            new_dedups[coin_id] = DedupCoinSpend(current_solution, spend_cost, duplicate_additions)
```

Comment:
it might make sense to keep this around in the `MempoolItem` as well. in case this won't make it into the block, we wouldn't want to have to run the puzzle again next time.